File: rosetta_source/GUIs/pyrosetta_toolkit/modules/DesignBreakdown.py
#!/usr/bin/python

# (c) Copyright Rosetta Commons Member Institutions.
# (c) This file is part of the Rosetta software suite and is made available under license.
# (c) The Rosetta software is developed by the contributing members of the Rosetta Commons.
# (c) For more information, see http://www.rosettacommons.org. Questions about this can be
# (c) addressed to University of Washington UW TechTransfer, email: license@u.washington.edu.

## @file   /GUIs/pyrosetta_toolkit/modules/DesignBreakdown.py
## @brief  Class for analyzing design results from a fasta of sequences
## @author Jared Adolf-Bryfogle (jadolfbr@gmail.com)


#Tkinter Imports
from Tkinter import *

#Python Imports
import re

#Toolkit Imports
from prettytable import prettytable
from PDB import *

class DesignBreakdown:
    """
    This class functions in organizing breakdowns for design.  Perhaps you can do this using features.  Probably.  Anyhow, this will work for the GUI.
    It can output a text file, a database, and run an R script to graph the results.
    It also has some GUI functions, which is why it's here.
    """
    
    def __init__(self, fasta_path):
        self.total_sequences = 0
        self.sequences = []
        self.frequency_map = dict()
        self.load_sequences(self, fasta_path)
        self.region = ""
        # Native PDB?
        
    def load_sequences(self, fasta_path):
        """
        Opens Fasta file and appends a list of sequences.
        """
        
        FILE = open(fasta_path, 'r')
        for line in FILE:
            if re.search(">", line):
                continue
            line = line.strip()
            self.sequences.append(line)
            
    def calculate_frequencies(self):
        pass
    
    def output_prettytable(self):
        pass
    
    def output_database(self):
        pass
    
    def run_R_script(self):
        pass
    
    def run_GUI_dialog(self):
        """
        Runs a series of GUI dialogs to get information from the user
        """
        pass

File: source/src/python/bindings/app/pyrosetta_toolkit/modules/R_Scripts/DesignBreakdown.R
#!/usr/bin/Rscript

# (c) Copyright Rosetta Commons Member Institutions.
# (c) This file is part of the Rosetta software suite and is made available under license.
# (c) The Rosetta software is developed by the contributing members of the Rosetta Commons.
# (c) For more information, see http://www.rosettacommons.org. Questions about this can be
# (c) addressed to University of Washington UW TechTransfer, email: license@u.washington.edu.

## @file   /GUIs/pyrosetta_toolkit/modules/R_Scripts/BreakdownDesign.R
## @brief  R Script for plotting design breakdown for modules/DesignBreakdown.py
## @author Jared Adolf-Bryfogle (jadolfbr@gmail.com)

#Args are: Database_In, Output_directory

#Database Columns Required: 
 #region, type (design or reference), rosetta_position, pdb_position(ex: 24L), aa, prob

#Requirements:
require(RSQLite)
require(rgl)


args=(commandArgs(TRUE))
if(length(args)==0){
    print("No arguments supplied.")}

if (length(args)!=2){
	print("Output Directory Required")}


drv <- dbDriver("SQLite")

con <- dbConnect(drv, dbname = toString(args[1]))
outdir = toString(args[2])

sql <- "SELECT DISTINCT aa FROM design_data WHERE type='design'"
aminos = dbGetQuery(con, sql)

sql <- "SELECT DISTINCT region FROM design_data"
regions = dbGetQuery(con, sql)

#sink(paste(sep="", outdir, "/", "LOG.txt"))
for (r in regions$region){
	f = paste(sep="", outdir,"/", r, "_PLOTS.pdf")
	if (regexpr(':', r)){
		sp = strsplit(r, ':')
		new = paste(sep="_", sp[[1]][1], sp[[1]][2], sp[[1]][3])
		f = paste(sep="", outdir,"/", new, "_PLOTS.pdf")
	}
	
	pdf(file=f, width=7*5/2, height=7*4/2)
	par(mfrow=c(4,3))


	sql <-"SELECT DISTINCT pdb_position FROM design_data WHERE region=?"
	pdb_positions = dbGetQuery(con, sql, r)

	for (position in pdb_positions$pdb_position){

		sql<-"SELECT prob FROM design_data WHERE type='design' and region=? and pdb_position=?"
		df = data.frame(region=r, pdb_position=position)
		prob = dbGetQuery(con, sql, df)

		sql<-"SELECT aa FROM design_data WHERE type='reference' and region=? and pdb_position=?"
		reference_aa = dbGetQuery(con, sql, df)

		aa = reference_aa$aa[1]
		
		#I think blue looks great.  If you want something else, change it here.
		barplot(prob$prob, names.arg=aminos$aa, density=prob$prob*100, col='blue')
		title(main=position, xlab=paste('AA (reference: ',aa, ')'), ylab='Percent Observed', font.main=2, cex.main=1.6, cex.lab=1.2)
		#legend('topright', legend = c(paste('reference:', aa)), box.col='blue', text.col='blue')
		
		

	}
	dev.off()
}
warnings()
#sink()
#close(con)


